lib/semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,385 @@
+    return selectCaseU16StringVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::u32string>::value,
+      std::map<std::u32string, parser::CharBlock> &>::type
+  GetValHandle() {
+    return selectCaseU32StringVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::int64_t>::value,
+      std::vector<SelectCaseRangeType<std::int64_t>> &>::type
+  GetRangeValHandle() {
+    return selectCaseIntRangeVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::string>::value,
+      std::vector<SelectCaseRangeType<std::string>> &>::type
+  GetRangeValHandle() {
+    return selectCaseStringRangeVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::u16string>::value,
+      std::vector<SelectCaseRangeType<std::u16string>> &>::type
+  GetRangeValHandle() {
+    return selectCaseU16StringRangeVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::u32string>::value,
+      std::vector<SelectCaseRangeType<std::u32string>> &>::type
+  GetRangeValHandle() {
+    return selectCaseU32StringRangeVals_;
+  }
+
+  std::map<bool, parser::CharBlock> selectCaseLogicalVals_;
+  std::map<std::int64_t, parser::CharBlock> selectCaseIntVals_;
+  std::map<std::string, parser::CharBlock> selectCaseStringVals_;
+  std::map<std::u16string, parser::CharBlock> selectCaseU16StringVals_;
+  std::map<std::u32string, parser::CharBlock> selectCaseU32StringVals_;
+  std::vector<SelectCaseRangeType<std::int64_t>> selectCaseIntRangeVals_;
+  std::vector<SelectCaseRangeType<std::string>> selectCaseStringRangeVals_;
+  std::vector<SelectCaseRangeType<std::u16string>>
+      selectCaseU16StringRangeVals_;
+  std::vector<SelectCaseRangeType<std::u32string>>
+      selectCaseU32StringRangeVals_;
+  int kind_;
+  TypeCategory category_;
+  parser::Messages &messages_;
+};
+
+void SelectStmtChecker::Leave(
+    const parser::CaseConstruct &selectCaseConstruct) {
+  const auto &selectCaseStmt{
+      std::get<parser::Statement<parser::SelectCaseStmt>>(
+          selectCaseConstruct.t)};
+  const auto &parsedExpr{
+      std::get<parser::Scalar<parser::Expr>>(selectCaseStmt.statement.t).thing};
+  std::optional<evaluate::DynamicType> selectCaseStmtType;
+  std::optional<TypeCategory> validCaseStmtTypeCategory;
+  if (const auto *expr{GetExpr(parsedExpr)}) {
+    if (auto type{expr->GetType()}) {
+      selectCaseStmtType = type;
+    }
+  }
+  if (!selectCaseStmtType) {
+    return;
+  }
+  if ((selectCaseStmtType.value().category() != TypeCategory::Integer) &&
+      (selectCaseStmtType.value().category() != TypeCategory::Character) &&
+      (selectCaseStmtType.value().category() !=
+          TypeCategory::Logical)) {  // C1145
+    context_.Say(parsedExpr.source,
+        "SELECT CASE expression must be of type CHARACTER, INTEGER, OR LOGICAL"_err_en_US);
+  } else {
+    validCaseStmtTypeCategory = selectCaseStmtType.value().category();
+  }
```

Comment:
This check does not seem to be handling if the select case expression is a derived type.
```
program mn
  type scores
    integer :: val
  end type
  type (scores) :: score
  select case (score)
  case(scores(100))
     print *, "century"
  end select
end
```

Reply:
Derived types are not allowed in select Case:
f18 throws below error as of now: 

test2.f90:25:16: error: SELECT CASE expression must be of type CHARACTER, INTERGER, OR LOGICAL
    select case (score)
                 ^^^^^

Reply:
Sorry about that. Yes, it does catch that error. But because execution continues it goes forward and hits an assert/check on selectCaseStmtType->kind() at 313.
You might not see this on a release build.

Reply:
I didnt notice that. Thanks for pointing this out Kiran. I will avoid extracting the kind if there is a mismatch in type.